### DISCUSSION TITLE
Rotating proxies and ratelimit mitigation

### DISCUSSION THREAD

**Discussion by nordkorea** [searxng/searxng#2567] [Q&A]
As far as I know, the searx.be instance hosted by ~either~ @unixfox ~or @return42~ is using rotating proxies to avoid ratelimiting and the search engines blocking "suspicious requests", how does it all work, how can I implement it into my own instance and what are more things I need to know about to make things as optimal and stable as possible?

**Comment by unixfox** [comment]
I don't really have rotating proxies. For some engines I pass the traffic through tor.

That's very easy to do, there is documentation available for that. It comes down to configuring a proxy in searxng.

**Comment by extremelyonline** [comment]
I use rotating proxies in my instance opnxng.com. The documentation on how to configure the settings.yml is clear.

One method is to set up SSH tunnels to another vps using the -D option. You can look into autossh (with systemd) to make the tunnel more permanent. After that, you should be able to add the socks5h:// proxy to the settings.yml. Other socks5 servers, such as [this one](https://github.com/serjs/socks5-server), also work.

I suppose one can also purchase proxies online, but I cannot speak to the privacy aspect of this method.

One thing to note is that, whenever a proxy fails, Searxng will be down and return proxy error. Another is that you can use proxy only for specific engines if you want to by configuring the settings.yml.

**I do have one question: when proxies are implemented, is the host ip included in the rotation?**

**Reply by dalf** [reply]
> One thing to note is that, whenever a proxy fails, Searxng will be down and return proxy error.

You can set `retries` to 2 or more, so SearXNG is going to try another proxy.

Currently, there is no rate limiting ( #1273 uses uwsgi, but we have dropped the dependency to it, it should use Redis).

> I do have one question: when proxies are implemented, is the host ip included in the rotation?

I'm not sure to understand your question.

Each new request is sent to the different proxies in turn using a [cycle](https://docs.python.org/3/library/itertools.html#itertools.cycle).
There is no health check, and the engine status is for the whole engine not for one specific proxy. I have some code to improve that, but I can't promise it is going to be released anytime soon..

Related code:
https://github.com/searxng/searxng/blob/7052a1a7cfa9aa691d854a9513d023e3fdc28fcf/searx/network/network.py#L145-L151

(Note: `pattern` is most probably equal to `all://` in your case)

**Reply by extremelyonline** [reply]
I see, thank you! Say I host Searxng on VPS A, and add VPS B as a proxy:

```
  proxies:
   all://:
      - socks5h://ip-of-VPS-B:8080
```

Does this mean that the ip of VPS A would not be used to make search requests?

(In my set up, I actually run a socks5 server on VPS A and included that as a "proxy". Was just wondering if that is redundant)

**Reply by dalf** [reply]
> Does this mean that the ip of VPS A would not be used to make search requests?

Yes, with `proxies` in the network configuration, only the defined proxies are used and nothing else, not even the clear connexion.
